src/proteus/session/Session.coffee

```diff
@@ -189,34 +188,30 @@ module.exports = class Session
       switch
         when msg instanceof CipherMessage
           return resolve @_decrypt_cipher_message envelope, envelope.message
-
         when msg instanceof PreKeyMessage
-          if msg.identity_key.fingerprint() != @remote_identity.fingerprint()
-            throw new DecryptError.RemoteIdentityChanged
-
-          decrypt_error = null
-
-          try
-            return resolve @_decrypt_cipher_message envelope, msg.message
-          catch e
-            decrypt_error = e
-
-            if not (e instanceof DecryptError.InvalidSignature or
-                    e instanceof DecryptError.InvalidMessage)
-              throw e
-
-          @_new_state prekey_store, msg
-          .then (state) =>
-            plaintext = state.decrypt envelope, msg.message
-            if msg.prekey_id != PreKey.MAX_PREKEY_ID
-              prekey_store.remove(msg.prekey_id)
-
-            @_insert_session_state msg.message.session_tag, state
-            @pending_prekey = null
-
-            return resolve plaintext
-          .catch =>
-            throw decrypt_error
+          throw new DecryptError.RemoteIdentityChanged if msg.identity_key.fingerprint() isnt @remote_identity.fingerprint()
+          return resolve @_decrypt_prekey_message envelope, msg, prekey_store
+        else
```

Comment:
We should get rid of this `new Promise` pattern since its error-prone. see https://jsfiddle.net/ybmLrzko/